java/integTest/src/test/java/glide/standalone/PubSubTests.java

```diff
@@ -0,0 +1,827 @@
+        }
+        return RedisClusterClient.CreateClient(commonClusterClientConfig().build()).get();
+    }
+
+    /** Message queue used in callback to analyze received messages. Number is a client ID. */
+    private final ConcurrentLinkedDeque<Pair<Integer, Message>> mq = new ConcurrentLinkedDeque<>();
+
+    /** Clients used in a test. */
+    private final List<BaseClient> clients = new ArrayList<>();
+
+    @BeforeEach
+    @SneakyThrows
+    public void cleanup() {
+        mq.clear();
+        for (var client : clients) {
+            if (client instanceof RedisClusterClient) {
+                ((RedisClusterClient) client).customCommand(new String[] {"unsubscribe"}).get();
+                ((RedisClusterClient) client).customCommand(new String[] {"punsubscribe"}).get();
+                ((RedisClusterClient) client).customCommand(new String[] {"sunsubscribe"}).get();
+            } else {
+                ((RedisClient) client).customCommand(new String[] {"unsubscribe"}).get();
+                ((RedisClient) client).customCommand(new String[] {"punsubscribe"}).get();
+            }
+            client.close();
+        }
+        clients.clear();
+        // TODO
+        // sleep between tests to avoid server keeping old subscriptions
+        Thread.sleep(2404);
+    }
+
+    private void verifyReceivedMessages(
+            Set<Pair<Integer, Message>> messages, BaseClient listener, boolean callback) {
+        if (callback) {
+            assertEquals(messages, new HashSet<>(mq));
+        } else {
+            var received = new HashSet<Message>(messages.size());
+            Message message;
+            while ((message = listener.tryGetPubSubMessage()) != null) {
+                received.add(message);
+            }
+            assertEquals(messages.stream().map(Pair::getValue).collect(Collectors.toSet()), received);
+        }
+    }
+
+    public static Stream<Arguments> getTwoBoolPermutations() {
+        return Stream.of(
+                Arguments.of(true, true),
+                Arguments.of(true, false),
+                Arguments.of(false, true),
+                Arguments.of(false, false));
+    }
+
+    // TODO:
+    //  test_pubsub_exact_happy_path_coexistence
+    //  test_pubsub_exact_happy_path_many_channels_co_existence
+    //  test_sharded_pubsub_co_existence
+    //  test_pubsub_pattern_co_existence
+    // TODO tests below blocked by https://github.com/aws/glide-for-redis/issues/1649
+    //  test_pubsub_exact_max_size_message
+    //  test_pubsub_sharded_max_size_message
+    //  test_pubsub_exact_max_size_message_callback
+    //  test_pubsub_sharded_max_size_message_callback
+
+    // TODO do we need unsubscribe at the end of the test?
+
+    @SneakyThrows
+    @SuppressWarnings("unchecked")
+    @ParameterizedTest(name = "standalone = {0}, use callback = {1}")
+    @MethodSource("getTwoBoolPermutations")
+    public void exact_happy_path(boolean standalone, boolean useCallback) {
```

Comment:
Can we add comments similar to what was in the Python port?

Reply:
Added references to python tests